add "ca_file" option to the elasticstack provider

utilitynerd opened the pull request:
allows connecting to an elasticsearch cluster using a custom certificate authority.

I know just enough go to get myself in trouble, but figured I would try to solve my own problem anyways. resolves #34 

It adds the "ca_file" parameter to the elasticstack provider.  If ca_file is set, the certificate is read and its contents are added to the CACert field in the elasticsearch.Config struct.

elasticmachine commented:
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?

Kushmaro commented:
Thanks so much for putting up the first community PR @utilitynerd , I have to ask though - is there a specific reason for not implementing "insecure=true" like mentioned in the issue? that would be the direction we'd like to take to resolve this one.

Specifying a root_ca file is often quite cumbersome, and doesn't add much value in TF as it does on browsers for example (where browsers will show if a cert is trusted or not)

olksdr commented on internal/clients/api_client.go:
```diff
@@ -82,6 +83,17 @@ func NewApiClientFunc(version string, p *schema.Provider) func(context.Context,
+							Severity: diag.Error,
+							Summary:  "Unable to read CA File",
+							Detail:   err.Error(),
+						})
```
```suggestion
						})
						return nil, diags
```
It would be great if we immediately return the `diags` here to report the error sooner.

olksdr reviewed (changes requested):
Hey @utilitynerd, 
Thanks for your contribution! 

There are few more steps to get this over and get merged:
* things we're also support the per elasticsearch-resource `elasticsearch_connection` block, we will need to add `ca_file` in there as well, see https://github.com/elastic/terraform-provider-elasticstack/blob/f0c6a1fe04b0883dbe3386087c58f20530a86d70/internal/utils/utils.go#L124-L149 so this way we will stay consistent in all the places, and then you can update code in  https://github.com/elastic/terraform-provider-elasticstack/blob/f0c6a1fe04b0883dbe3386087c58f20530a86d70/internal/clients/api_client.go#L123-L135 to read the newly added option
* after changes are maybe, the docs must be updates as well, you can run `make gen` command which will re-generate the docs and adds this options as well
* and one more step is you will have you sign the [Contributor Agreement](https://www.elastic.co/contributor-agreement) for the account you're pushing your commits from,  in this case it looks like the commit is coming from https://github.com/mikejones-ucb

Once this is done we can once more go over the PR and see if there are any missing things to be added.

Thanks again for your contribution!

utilitynerd commented:
Thanks for the feedback!

  I'm not opposed to the "insecure=true" method instead of validating against a custom cert.  Although maybe it would be nice to eventually have both options available.

Since it sounds like the project would prefer the insecure route, why don't I take your suggestions and see if I can  put together a separate PR implementing it.

utilitynerd commented:
I believe I've addressed the issues above:
- return early loading the ca_file results in an error
- added the ca_file parameter to the elasticsearch_connection block 
- generated updated documentation
- signed the contributor agreement with my other username 


Let me know if there is anything else to address

olksdr commented:
jenkins test this please

olksdr commented:
@utilitynerd please, merge `main` into your branch and resolve the conflicts which are coming from #36

olksdr commented:
ok to test

olksdr reviewed (approved):
lgtm
